### ISSUE TITLE
Redirect to callback route after SAML request fails

### ISSUE THREAD

**Issue by mattc-eostar** [auth0/auth0-react#250]
By submitting an Issue to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem

Thank you for troubleshooting [this](https://github.com/auth0/auth0-react/issues/145) issue. Setting up an open /callback route fixed one of the issues I was having but only when logging in directly to the web app.

I need this same behavior with SAML SSO. So the redirect URI is *.auth0.com/samlp/..... with a RelayState parameter that points to ```encodeURIComponent(`${window.location.origin}/callback`)```

The authentication of the third-party app is successful and it properly redirects back into my app to the callback route, but no `code` or `state` parameters come with it and it just hangs on /callback. I am assuming those parameters are getting eaten in a prior step, but should they be passed along?

Perhaps, I misunderstand how this should be working. I added `https://localhost:3000/callback` and `https://[AUTH0_DOMAIN].auth0.com/samlp/.....?RelayState=https://localhost:3000/callback` to my allowed callback URLs along with setting up the SAML 2 Web App addon. Again, the authentication works with that third-party app, just not coming back to my app that uses some authenticated iframes and components from that third-party, thus needing the authentication.

I am using `onRedirectCallback()` defined as:
```
  const onRedirectCallback = (appState) => {
    props.history.push(appState?.returnTo || window.location.pathname);
  };
```
Again, this all works properly when the redirect URI is `${window.location.origin}/callback`, but then no third-party authentication takes place.

Are there other required steps when doing this with SAML requests?

### What was the expected behavior?

Expected behavior would be:
1. Unauthenticated user comes to `/` route and redirected to universal login
2. User logs in
3. Auth0 makes SAML request to third party app with a relay state to return my app
4. Returns to `/callback` with code and state parameters and authenticates on my app
5. Redirects to index route fully authenticated

### Reproduction

```
const ProtectedRoute = ({ component, ...args }) => (
  <Route
    component={withAuthenticationRequired(component)}
    {...args}
  />
);

const AppRoutes = () => {
    return (
        <div className={styles.AppRouter}>
            <Switch>
                <ProtectedRoute exact path="/" component={Home} />
                <Route path="/callback" exact />
            </Switch>
        </div>
    );
};

const Auth0ProviderWithHistory = (props) => {
  const domain = process.env.REACT_APP_AUTH0_DOMAIN;
  const clientId = process.env.REACT_APP_AUTH0_CLIENT_ID;

  const onRedirectCallback = (appState) => {
    props.history.push(appState?.returnTo || window.location.pathname);
  };

  const relayState = encodeURIComponent(`${window.location.origin}/callback`);
  const redirectUri = `https://[AUTH0_DOMAIN].auth0.com/samlp/...?RelayState=${relayState}`;

  return (
    <Auth0Provider
      domain={domain}
      clientId={clientId}
      // redirectUri="https://localhost:3000/callback"
      redirectUri={redirectUri}
      onRedirectCallback={onRedirectCallback}
    >
      {props.children}
    </Auth0Provider>
  );
};

function App() {
    return (
        <Router>
            <Auth0ProviderWithHistory>
                <AppRoutes/>
            </Auth0ProviderWithHistory>
        </Router>
        
    );
}
```

### Environment

> Please provide the following:

- **Version of `auth0-react` used:** 1.5.0
- **Which browsers have you tested in?** Chrome
- **Which framework are you using, if applicable (Angular, React, etc):** React
- **Other modules/plugins/libraries that might be involved:** react-router-dom 5.0.1

**Comment by frederikprijck** [comment]
Hey,

Thanks for reporting this.
We are looking into this, but will need to sync with some of our more SAML experts as it's not a common thing with our SPA SDKs.

Will get back to this when we have more information.

Thanks.

**Comment by mattc-eostar** [comment]
Thank you! I am not even sure if this is happening on the Auth0 side or if it is happening on the third-party service side. Trying to figure that out.

In the meantime, it appears that I can get something working where I set the redirectUri to my callback route. Then in the OnRedirectCallback, I call `window.location.replace("https://[AUTH0_DOMAIN].auth0.com/samlp/...?RelayState=https://localhost:3000")` and it completes the loop by directing the browser to the samlp endpoint after authenticating a user for my app through Auth0.

Not sure if this is a bad idea, but the alternative (as recommended by the third-party service we are trying to integrate with) is to prepend the samlp endpoint anytime we need to use an asset in their domain to make sure the user is authenticated. We are loading dashboards in an iframe. This causes a lot of redundant calls to authenticate though so the OnRedirectCallback seems to get around that for now.

**Comment by frederikprijck** [comment]
Sorry for going a bit silent, but we had to also do some research here. 

The SDK being used here, `Auth0-React`, is built on top of our [`Auth0-SPA-JS` SDK ](https://github.com/auth0/auth0-spa-js). That latter SDK only supports OAuth2's Authorization Code flow using PKCE and OpenId Connect.

The idea would be to log in to Auth0 using OpenId Connect and keep the SAML stuff outside of this SDK, most likely by doing what the 3rd party provider is recommending.

I would advise asking on https://community.auth0.com/, there might be people there that are more familiar with SAML. However, as the React and SPA JS SDK's do not support SAML, this repository might not be the most effective way to get assistance.

Closing this, but feel free to ping if there is anything we could still (try and) assist with.
